label_studio/users/functions/last_activity.py

```diff
@@ -0,0 +1,560 @@
+        return {'success': True, 'processed': 0, 'errors': 0}
+
+    processed = 0
+    errors = 0
+
+    try:
+        with transaction.atomic():
+            # Get existing users
+            User = get_user_model()
+            user_ids = [activity['user_id'] for activity in activities]
+            existing_users = User.objects.filter(id__in=user_ids).only('id', 'last_activity')
+            user_dict = {user.id: user for user in existing_users}
+
+            # Update activities
+            users_to_update = []
+
+            for activity in activities:
+                user_id = activity['user_id']
+                new_activity = activity['last_activity']
+
+                user = user_dict.get(user_id)
+                if user:
+                    # Only update if the new activity is more recent
+                    if user.last_activity is None or new_activity > user.last_activity:
+                        user.last_activity = new_activity
+                        users_to_update.append(user)
+                        processed += 1
+                    else:
+                        logger.debug(
+                            'Skipping outdated activity for user %s: %s <= %s',
+                            user_id,
+                            new_activity,
+                            user.last_activity,
+                        )
+                        processed += 1
+                else:
+                    logger.warning('User %s not found in database', user_id)
+                    errors += 1
+
+            # Bulk update
+            if users_to_update:
+                User.objects.bulk_update(users_to_update, ['last_activity'], batch_size=100)
+                logger.info('Bulk updated %s users', len(users_to_update))
+
+            return {'success': True, 'processed': processed, 'errors': errors, 'updated': len(users_to_update)}
+
+    except Exception as e:
+        logger.error('Failed to bulk update user activities: %s', e, exc_info=True)
+        return {
+            'success': False,
+            'processed': processed,
+            'errors': errors + 1,
+            'message': f'Bulk update failed: {str(e)}',
+        }
+
+
+def schedule_activity_sync(force: bool = False) -> bool:
+    """
+    Schedule user activity synchronization if needed.
+
+    Args:
+        force: Force sync even if threshold not reached
+
+    Returns:
+        True if sync was scheduled, False otherwise
+    """
+    if not force and not should_sync_activities():
+        logger.debug('Sync threshold not reached, skipping')
+        return False
+
+    try:
+        # Schedule the sync job
+        start_job_async_or_sync(sync_user_activities_to_db, queue_name='low', redis=True)
+
+        logger.info('Scheduled user activity sync job')
+        return True
+
+    except Exception as e:
+        logger.error('Failed to schedule activity sync: %s', e)
+        return False
+
+
+@job('low', timeout=7200)
+def full_activity_sync() -> dict:
```

Comment:
This is a manually invoked task?

Reply:
I used it for testing, I'm not going to merge it.